### PR TITLE
boost up smth

### PR DESCRIPTION
**Pull request by rakkit**
i fucked my git history. You can only check out  https://github.com/janEbert/torchtitan/commit/9ddd725f25e5b9e776769a735d9165f0a2d55eeb and other commits before that are synced to your latest streams.  

i will write comment on each files

### CODE REVIEW

File: torchtitan/experiments/kernels/FBGEMM_moe/utils.py
Comment (rakkit) [inline torchtitan/experiments/kernels/FBGEMM_moe/utils.py:1]: this kernels from [fbgemm_gpu](https://github.com/pytorch/FBGEMM/blob/main/fbgemm_gpu/experimental/gemm/triton_gemm/utils.py)

(BSD license)

File: torchtitan/experiments/kernels/FBGEMM_moe/grouped_gemm.py
Comment (rakkit) [inline torchtitan/experiments/kernels/FBGEMM_moe/grouped_gemm.py:1]: this kernels from [fbgemm_gpu](https://github.com/pytorch/FBGEMM/blob/main/fbgemm_gpu/experimental/gemm/triton_gemm/utils.py)

(BSD license)

File: torchtitan/experiments/kernels/moe/indices.py
Comment (rakkit) [inline torchtitan/experiments/kernels/moe/indices.py:1]: this i manually sync the lastest version in torchtitan

[Commit 7183671](https://github.com/pytorch/torchtitan/commit/7183671742c5b688d184dc6240ec8f8105799e83)

File: torchtitan/models/MoEllama/GroupedExperts.py
Comment (rakkit) [inline torchtitan/models/MoEllama/GroupedExperts.py:1]: previous BMM, now grouped GeMM.

to use FBGEMM's kernel. The weights should be organised as `(experts dim_out) din_in` so that we need to reshape the weights before calling GMM.

Also moved the `.to_local()` in the weights reshape function to simplified code in forward. though we should disable EP for now

File: torchtitan/models/MoEllama/__init__.py
Comment (rakkit) [inline torchtitan/models/MoEllama/__init__.py:1]: I forgot to remove it.

easy for my local debugging. You can keep it in case anyone wants to debug on A100 or just ignore it

File: torchtitan/models/MoEllama/moellama.py
Comment (rakkit) [inline torchtitan/models/MoEllama/moellama.py:1]: refactor a bit.

1. Now we only do EP=1
2. add `@torch.compile(fullgraph=True)` manually for `Gate` and `loss` as these two parts do not bothered by dynamic graph 
3. move `routed_scaling_factor` from `Gate` to `MoE`
4. Now we `sum` experts' output into `fp32`. see `REMIX_DTYPE`. This does not cost too much memory, though I don't know if we can get any benefits there.  -> We might also need to manually set Gate's weights to FP32 in the FSDP wrapper. 
5. others refacors to make GEMM works and use permute kernel from `torchtitan`

File: torchtitan/models/MoEllama/parallelize_MoEllama.py [hunk 1]
```diff
@@ -416,7 +423,13 @@ def apply_compile(model: nn.Module):
     repeated structure. Alternatively one can compile the whole model (after applying DP).
     """
     for layer_id, transformer_block in model.layers.named_children():
-        transformer_block = torch.compile(transformer_block, dynamic=True)
+        # transformer_block = torch.compile(transformer_block, dynamic=True)
```
Comment (rakkit) [inline torchtitan/models/MoEllama/parallelize_MoEllama.py:426]: others in this file are formating. beblow is the change ->  we set `dynamic=False` for compile and set some strange `dynamo` flags to make compile work with new triton kernels.

File: torchtitan/models/llama3/model.py
Comment (rakkit) [inline torchtitan/models/llama3/model.py:1]: 1. Formating
2. add `@torch.compile(fullgraph=True)` for Attention's forward for MoE.   Don't want to compile the attention model in parallel_model.py,    but don't if this flag will broken the checkpoint or not yet

File: torchtitan/train.py [hunk 2]
```diff
@@ -17,6 +17,7 @@
 import torch
 from torch.distributed.elastic.multiprocessing.errors import record
 
+import torchtitan.models
```
Comment (rakkit) [inline torchtitan/train.py:20]: 1. Both of us need this to make `sbatch` work, no clue why yet

2. Raise an error when EP>1 is used.

3. formating

File: torchtitan/experiments/kernels/moe/indices.py
Comment (janEbert) [reply torchtitan/experiments/kernels/moe/indices.py:1]: Thanks for documenting this. Very helpful, I was wondering what the change was about. :)
